packages/SwingSet/src/kernel/kernelSyscall.js

```diff
@@ -100,58 +100,59 @@ export function makeKernelSyscallHandler(tools) {
     return OKNULL;
   }
 
-  function fulfillToPresence(vatID, kpid, targetSlot) {
-    insistVatID(vatID);
-    insistKernelType('promise', kpid);
-    insistKernelType('object', targetSlot);
-    kernelKeeper.incStat('syscalls');
-    kernelKeeper.incStat('syscallFulfillToPresence');
-    const p = kernelKeeper.getResolveablePromise(kpid, vatID);
-    const { subscribers, queue } = p;
-    kernelKeeper.fulfillKernelPromiseToPresence(kpid, targetSlot);
-    notifySubscribersAndQueue(kpid, vatID, subscribers, queue);
-    // todo: some day it'd be nice to delete the promise table entry now. To
-    // do that correctly, we must make sure no vats still hold pointers to
-    // it, which means vats must drop their refs when they get notified about
-    // the resolution ("you knew it was resolved, you shouldn't be sending
-    // any more messages to it, send them to the resolution instead"), and we
-    // must wait for those notifications to be delivered.
-    if (p.policy === 'logAlways') {
-      console.log(`${kpid}.policy logAlways: fulfillToPresence ${targetSlot}`);
+  function extractPresenceIfPresent(data) {
+    const body = JSON.parse(data.body);
+    if (
+      body &&
+      typeof body === 'object' &&
+      body['@qclass'] === 'slot' &&
+      body.index === 0
+    ) {
+      if (data.slots.length === 1) {
+        const slot = data.slots[0];
+        const { type } = parseKernelSlot(slot);
+        if (type === 'object') {
+          return slot;
+        }
+      }
     }
-    return OKNULL;
+    return null;
   }
 
-  function fulfillToData(vatID, kpid, data) {
+  function resolve(vatID, resolutions) {
     insistVatID(vatID);
-    insistKernelType('promise', kpid);
-    insistCapData(data);
     kernelKeeper.incStat('syscalls');
-    kernelKeeper.incStat('syscallFulfillToData');
-    const p = kernelKeeper.getResolveablePromise(kpid, vatID);
-    const { subscribers, queue } = p;
-    let idx = 0;
-    for (const dataSlot of data.slots) {
-      kernelKeeper.incrementRefCount(dataSlot, `fulfill|s${idx}`);
-      idx += 1;
+    kernelKeeper.incStat('syscallResolve');
+    for (const resolution of resolutions) {
+      const [kpid, rejected, data] = resolution;
+      insistKernelType('promise', kpid);
+      insistCapData(data);
+      if (rejected) {
+        resolveToError(kpid, data, vatID);
+      } else {
+        const p = kernelKeeper.getResolveablePromise(kpid, vatID);
+        const { subscribers, queue } = p;
+        let idx = 0;
+        for (const dataSlot of data.slots) {
+          kernelKeeper.incrementRefCount(dataSlot, `resolve|s${idx}`);
```

Comment:
No, that's not what's happening here.  We're not recording which slot points to something, we're just refcounting.  `dataSlot` is not a slot index, it's the value in the slot, i.e., the `kpid` of the referenced promise whose reference count needs incrementing because the slot entry references it.